### PR TITLE
Node electron build

### LINKED ISSUE #6138: Error running gRPC greeter example with electron
Running the node gRPC greeter example with electron will throw the following error. 
I'm using [electron](http://electron.atom.io/) v0.37.5 which is using Node 5.10.0.
I'm not sure if this error should be fixed inside gRPC or electron. If it should be fixed inside electron, maybe someone could give some details how to fix this, so I can file a bug against the electron project.

[Error: /projects/grpc/examples/node/node_modules/grpc/src/node/extension_binary/grpc_node.node: undefined symbol: GENERAL_NAME_free]
A JavaScript error occurred in the main process
Uncaught Exception:
Error: 
/projects/grpc/examples/node/node_modules/grpc/src/node/extension_binary/grpc_node.node: undefined symbol: GENERAL_NAME_free
    at Error (native)
    at process.module.(anonymous function) [as dlopen](ATOM_SHELL_ASAR.js:158:20)
    at Object.Module._extensions..node (module.js:440:18)
    at Object.module.(anonymous function) [as .node](ATOM_SHELL_ASAR.js:158:20)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (/projects/grpc/examples/node/node_modules/grpc/src/node/src/grpc_extension.js:38:15)

### PR DESCRIPTION
**Pull request by murgatroid99**
This allows the Node library to be built for Electron by following [the existing procedure on the Electron website](http://electron.atom.io/docs/tutorial/using-native-node-modules/). It also adds scripts for building artifacts for Electron and running the tests on Electron. This fixes #6138 and maybe #8166.

Electron does not work with the libuv endpoint implementation, so we will continue to need the non-uv implementation. So, I also refactored the uv/non-uv split in the Node extension code.

This is built on #8588, so that should be merged first

### CODE REVIEW

**Comment by casesandberg** [comment]
YESSS 🐙

**Comment by casesandberg** [comment]
Tried building this locally with `npm run electron-build -- --target=6.7.0` and got:
```
../src/boringssl/err_data.c:17:10: fatal error: 'openssl/base.h' file not found
#include <openssl/base.h>
```

**Comment by murgatroid99** [comment]
When you say "locally", I assume you mean that you cloned the git repository. Did you update the submodules? (`git submodule update --init`)

**Comment by jcvernaleo** [comment]
Just going to chime in to say that I was able to build this locally and use it with electron (1.4.6, node 6.5.0).  Really hope it makes it in.

**Comment by alexlyp** [comment]
Was also able to use this build to work with electron.  Though ran into this issue when attempting to connect a client.

> electron .

E1111 12:24:34.582899941   13662 handshake.c:128]            Security handshake failed: {"created":"@1478888674.582671051","description":"Handshake read failed","file":"../src/core/lib/security/transport/handshake.c","file_line":237,"referenced_errors":[{"created":"@1478888674.582595366","description":"FD shutdown","file":"../src/core/lib/iomgr/ev_epoll_linux.c","file_line":1045}]}
{ Error: Connect Failed
    at /home/user/git/src/github.com/decred/decrediton/node_modules/grpc/src/node/src/client.js:417:17 code: 14, metadata: Metadata { _internal_repr: {} } }

The code itself is pretty simple.  Basic electron-quick-start app with this added on: 
https://github.com/decred/dcrwallet/blob/master/rpc/documentation/clientusage.md#nodejs

I verified that this example works without electron and just running with node

**Comment by murgatroid99** [comment]
OK, that looks like a TLS error, and it makes sense that that could be affected by this change, because gRPC uses a different TLS implementation under Electron. 

I'm guessing that the certificate is self-signed and uses an IP address for the server name. If so, that may be the same problem as issue #6722 (Node gRPC on Electron and Node gRPC on Windows both use BoringSSL instead of OpenSSL). If so, I think that would confirm the theory that that problem is related to BoringSSL.

Unfortunately, as things are currently set up, BoringSSL is the only viable option for Node gRPC on Electron, so that problem may not be fixed before we publish official Electron support.

**Comment by casesandberg** [comment]
I am getting a bunch of errors when trying to build after updating the submodules:

```
  CC(target) Release/obj.target/boringssl/src/boringssl/err_data.o
In file included from ../src/boringssl/err_data.c:18:
In file included from /Users/case/.node-gyp/6.7.0/include/node/openssl/err.h:115:
/Users/case/.node-gyp/6.7.0/include/node/openssl/e_os2.h:276:11: warning: 'OPENSSL_EXPORT' macro redefined [-Wmacro-redefined]
#  define OPENSSL_EXPORT extern
          ^
../third_party/boringssl/include/openssl/base.h:137:9: note: previous definition is here
#define OPENSSL_EXPORT
        ^
In file included from ../src/boringssl/err_data.c:18:
In file included from /Users/case/.node-gyp/6.7.0/include/node/openssl/err.h:122:
/Users/case/.node-gyp/6.7.0/include/node/openssl/ossl_typ.h:152:29: error: typedef redefinition with different types ('struct ecdsa_method' vs 'struct ecdsa_method_st')
typedef struct ecdsa_method ECDSA_METHOD;
                            ^
../third_party/boringssl/include/openssl/base.h:217:32: note: previous definition is here
typedef struct ecdsa_method_st ECDSA_METHOD;
                               ^
In file included from ../src/boringssl/err_data.c:18:
In file included from /Users/case/.node-gyp/6.7.0/include/node/openssl/err.h:124:
In file included from /Users/case/.node-gyp/6.7.0/include/node/openssl/bio.h:69:
In file included from /Users/case/.node-gyp/6.7.0/include/node/openssl/crypto.h:130:
/Users/case/.node-gyp/6.7.0/include/node/openssl/opensslv.h:33:10: warning: 'OPENSSL_VERSION_NUMBER' macro redefined [-Wmacro-redefined]
# define OPENSSL_VERSION_NUMBER  0x100020afL
         ^
../third_party/boringssl/include/openssl/base.h:112:9: note: previous definition is here
#define OPENSSL_VERSION_NUMBER 0x10002000
        ^
In file included from ../src/boringssl/err_data.c:18:
In file included from /Users/case/.node-gyp/6.7.0/include/node/openssl/err.h:124:
In file included from /Users/case/.node-gyp/6.7.0/include/node/openssl/bio.h:69:
/Users/case/.node-gyp/6.7.0/include/node/openssl/crypto.h:451:3: error: typedef redefinition with different types ('struct crypto_threadid_st' vs 'int')
} CRYPTO_THREADID;
  ^
../third_party/boringssl/include/openssl/base.h:151:13: note: previous definition is here
typedef int CRYPTO_THREADID;
            ^
In file included from ../src/boringssl/err_data.c:19:
../third_party/boringssl/include/openssl/type_check.h:77:9: warning: 'CHECKED_PTR_OF' macro redefined [-Wmacro-redefined]
#define CHECKED_PTR_OF(type, p) CHECKED_CAST(void*, type*, (p))
        ^
/Users/case/.node-gyp/6.7.0/include/node/openssl/safestack.h:65:11: note: previous definition is here
#  define CHECKED_PTR_OF(type, p) \
          ^
../src/boringssl/err_data.c:33:1: error: static_assert failed "library_values_changed_12"
OPENSSL_COMPILE_ASSERT(ERR_LIB_ASN1 == 12, library_values_changed_12);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:34:1: error: static_assert failed "library_values_changed_13"
OPENSSL_COMPILE_ASSERT(ERR_LIB_CONF == 13, library_values_changed_13);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:35:1: error: static_assert failed "library_values_changed_14"
OPENSSL_COMPILE_ASSERT(ERR_LIB_CRYPTO == 14, library_values_changed_14);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:36:1: error: static_assert failed "library_values_changed_15"
OPENSSL_COMPILE_ASSERT(ERR_LIB_EC == 15, library_values_changed_15);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:37:1: error: static_assert failed "library_values_changed_16"
OPENSSL_COMPILE_ASSERT(ERR_LIB_SSL == 16, library_values_changed_16);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:38:1: error: static_assert failed "library_values_changed_17"
OPENSSL_COMPILE_ASSERT(ERR_LIB_BIO == 17, library_values_changed_17);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:39:1: error: static_assert failed "library_values_changed_18"
OPENSSL_COMPILE_ASSERT(ERR_LIB_PKCS7 == 18, library_values_changed_18);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:40:24: error: use of undeclared identifier 'ERR_LIB_PKCS8'
OPENSSL_COMPILE_ASSERT(ERR_LIB_PKCS8 == 19, library_values_changed_19);
                       ^
../src/boringssl/err_data.c:41:1: error: static_assert failed "library_values_changed_20"
OPENSSL_COMPILE_ASSERT(ERR_LIB_X509V3 == 20, library_values_changed_20);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:42:1: error: static_assert failed "library_values_changed_21"
OPENSSL_COMPILE_ASSERT(ERR_LIB_RAND == 21, library_values_changed_21);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:43:1: error: static_assert failed "library_values_changed_22"
OPENSSL_COMPILE_ASSERT(ERR_LIB_ENGINE == 22, library_values_changed_22);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:44:1: error: static_assert failed "library_values_changed_23"
OPENSSL_COMPILE_ASSERT(ERR_LIB_OCSP == 23, library_values_changed_23);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:45:1: error: static_assert failed "library_values_changed_24"
OPENSSL_COMPILE_ASSERT(ERR_LIB_UI == 24, library_values_changed_24);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:46:1: error: static_assert failed "library_values_changed_25"
OPENSSL_COMPILE_ASSERT(ERR_LIB_COMP == 25, library_values_changed_25);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:47:1: error: static_assert failed "library_values_changed_26"
OPENSSL_COMPILE_ASSERT(ERR_LIB_ECDSA == 26, library_values_changed_26);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:48:1: error: static_assert failed "library_values_changed_27"
OPENSSL_COMPILE_ASSERT(ERR_LIB_ECDH == 27, library_values_changed_27);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
../src/boringssl/err_data.c:49:1: error: static_assert failed "library_values_changed_28"
OPENSSL_COMPILE_ASSERT(ERR_LIB_HMAC == 28, library_values_changed_28);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../third_party/boringssl/include/openssl/type_check.h:80:43: note: expanded from macro 'OPENSSL_COMPILE_ASSERT'
#define OPENSSL_COMPILE_ASSERT(cond, msg) _Static_assert(cond, #msg)
                                          ^              ~~~~
fatal error: too many errors emitted, stopping now [-ferror-limit=]
3 warnings and 20 errors generated.
```

**Comment by murgatroid99** [comment]
That error output has a bunch of `In file included from /Users/case/.node-gyp/6.7.0/include/node/openssl/err.h`. That looks like it's building against the current version of Node, instead of against Electron. Do you have all of the environment variables set correctly?

**Comment by casesandberg** [comment]
What ENV variables do I need set? I am just running `npm run electron-build -- --target=6.7.0`

**Comment by murgatroid99** [comment]
6.7.0 isn't a valid Electron target. The latest version of Electron is 1.4.6. And I would strongly recommend setting the environment variable `HOME=~/.electron-gyp`, as suggested in the Electron docs linked above, so that the downloaded files don't conflict with existing files.

**Comment by eltoro** [comment]
Is there any estimated time when this will be done? Thanks

**Comment by murgatroid99** [comment]
@eltoro Are you saying that this pull request doesn't work as it currently exists? If so, what problem did you see?

**Comment by eltoro** [comment]
@murgatroid99 I didn't try out the PR. I ended up just forking the repo and then just updating the binding.gyp. That worked for me. I'm on MacOS 10.12.1.

**Comment by jcvernaleo** [comment]
So I can build the electron friendly grpc with the command:
```
npm run electron-build -- --target=1.4.6
```
or even
```
./tools/run_tests/build_node_electron.sh 1.4.6
```
What I'm not seeing yet is how can I make a grpc tgz with that build that I can install with npm install on another box (or just in another directory)?  Is there some other npm script that I'm missing (or one of the many scripts in tools/)?

**Comment by murgatroid99** [comment]
It's not the tarball that's the problem. That's easy enough to get with `npm pack`. But that won't contain the compiled stuff. If you have the right tools set up, you should be able to install this the same way you would install any other native module using the instructions from Electron that I linked above.

**Comment by jcvernaleo** [comment]
I guess what I was hoping for was a way to make ```npm pack``` include the compiled stuff.  Since there doesn't seem to be a way I'll go back to the electron link.  Thanks.

**Comment by alexlyp** [comment]
Running into some really strange errs when babel-loader via webpack tries to build the grpc module.
ERROR in ./~/grpc/package.json
Module parse failed: /home/user/git/src/github.com/decred/decrediton/node_modules/grpc/package.json Unexpected token (2:9)
You may need an appropriate loader to handle this file type.
SyntaxError: Unexpected token (2:9)

ERROR in ./~/grpc/~/node-pre-gyp/lib/info.js
Module not found: Error: Cannot resolve module 'aws-sdk' in /home/user/git/src/github.com/decred/decrediton/node_modules/grpc/node_modules/node-pre-gyp/lib
 @ ./~/grpc/~/node-pre-gyp/lib/info.js 14:14-32

etc...

All the other installed modules appear to build fine, but grpc's can't seem to parse even the simplest of js or json.

**Comment by alexlyp** [comment]
I did notice some inexplicable issues when trying to use watchify + babelify with grpc as well.  It would bundle everything nicely, then a random '_handle' not defined somewhere in that 100K bundle.js 

Would go away when I commented out the import grpc

**Comment by murgatroid99** [comment]
I don't know what babel-loader does, so I don't know why it doesn't work. But the babel-loader page seems to indicate that you should be excluding `node_modules` from whatever it does, so you shouldn't be building your dependencies. It may be that your `file: "../grpc"` pointer is bypassing that.

**Comment by cristiancavalli** [comment]
@murgatroid99 any idea whether or not we can provide some instructions on back-porting this to the VS code team? I'm currently trying to create an extension which uses GRPC and the dll errors persist.

**Comment by mnaamani** [comment]
Thank you @murgatroid99 for this patch.

I was able to successfully build the module from source, for electron 1.4.11 on OSX Sierra, node v6.5.0

The grpc module is loaded successfully in the electron app and works as expected.

There is an error however when trying to run `npm test` .

Any ideas?

grpc@1.1.0-dev test /Users/dev/Electron.app/Contents/Resources/app/node_modules/grpc
mocha src/node/test && npm run-script lint

module.js:583
  return process.dlopen(module, path._makeLong(filename));
                 ^
Error: Module version mismatch. Expected 48, got 50.
    at Error (native)
    at Object.Module._extensions..node (module.js:583:18)
    at Module.load (module.js:473:32)
    at tryModuleLoad (module.js:432:12)
    at Function.Module._load (module.js:424:3)
    at Module.require (module.js:483:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/dev/Electron.app/Contents/Resources/app/node_modules/grpc/src/node/src/grpc_extension.js:38:15)

..... snip .....

npm ERR! Test failed.  See above for more details.

**Comment by murgatroid99** [comment]
Are you sure `npm test` is using Electron and not Node? Either way, that error indicates a version mismatch between the version of Electron that you built for and the version of Electron/Node that you're running.

**Comment by mnaamani** [comment]
I'm not sure as a matter of fact. What would be the right command line to run `npm test` using electron?

**Comment by murgatroid99** [comment]
It looks like you installed gRPC as a dependency for your own app, and you're running `npm test` for that app. If that's accurate, you should just do what you normally do to run your tests on Electron.

If you're asking how to run the gRPC tests in the gRPC repository using electron, you should be able to do something like this (assuming you have electron installed globally)

```sh
electron ./node_modules/.bin/electron-mocha src/node/test
```

**Comment by mnaamani** [comment]
Yes that is what I was looking for:

```
electron ./node_modules/.bin/electron-mocha src/node/test
```

All unit tests are passing :+1:

Hope to see this merged soon.

**Comment by murgatroid99** [comment]
test this please

**Comment by jcvernaleo** [comment]
@murgatroid99 we've been using this branch on both Linux and OSX for a week or two with no issue (just in case that helps).

**Comment by sadeghis** [comment]
Works like a charm! This is a great pull request, thank you @murgatroid99.

**Comment by casesandberg** [comment]
Is there an ETA on when this will be merged in?

**Comment by jcvernaleo** [comment]
@murgatroid99 Is there anything on the dev front you need to help get this in?  We've been using it for decred/decrediton and all seems well for us but of course I'd be happier with something in master (and would be more than happy to help with anything if needed).

**Comment by murgatroid99** [comment]
The test result links are screwed up. Jenkins: test this please

File: tools/dockerfile/test/node_jessie_x64/Dockerfile [hunk 1]
```diff
@@ -42,12 +42,18 @@ RUN apt-get update && apt-get install -y \
   git \
   golang \
   gyp \
+  libasound2 \
```
Comment (nicolasnoble) [inline tools/dockerfile/test/node_jessie_x64/Dockerfile]: Really ? :D

**Comment by murgatroid99** [comment]
The interop failure is #8989.

**Comment by murgatroid99** [comment]
The Linux failure is #9150. The TSAN failure is #9124. The Mac failure is #9241. All other tests pass, and in addition, the Node artifact build completes successfully (https://grpc-testing.appspot.com/view/Artifacts/job/gRPC_build_artifacts/2570/).

**Pull request merged**